mne/decoding/gsoc.py

```diff
@@ -0,0 +1,195 @@
+# Authors: Jean-Remi King <jeanremi.king@gmail.com>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+from .mixin import TransformerMixin
+from ..epochs import _BaseEpochs
+
+
+class EpochsTransformerMixin(TransformerMixin):
+    """Mixin class for reshaping data to Epoch's standard shape
+
+    This class is meant to be inherited by transformers meant to be
+    used in scikit-learn pipeline. Useful utility for making data
+    compatible with mne internal functions.
+
+    Parameters
+    ----------
+    n_chan : int | None
+
+        The number of channels.
+    """
+
+    def __init__(self, n_chan=None):
+        self.n_chan = n_chan
+        self._check_init()
+
+    def fit(self, X, y=None):
+        """No use here. Added for scikit-learn compatibility.
+
+        Parameters
+        ----------
+        X : numpy array of dimensions [2,3,4]
+            The data to be reshaped.
+        y : None
+            Used for scikit-learn compatibilty
+
+        Returns
+        -------
+        self : Instance of EpochsTransformerMixin
+            Return the same object.
+        """
+        return self
+
+    def fit_transform(self, X, y=None):
+        """No use here. Added for scikit-learn compatibility.
+
+        Parameters
+        ----------
+        X : numpy array of dimensions [2,3,4]
+            The data to be reshaped.
+        y : None
+            Used for scikit-learn compatibilty
+
+        Returns
+        -------
+        X : numpy ndarray
+            The same array
+        """
+        self.fit(X, y)
+        return self.transform(X)
+
+    def transform(self, X):
+        """No use here. Added for scikit-learn compatibility.
+
+        Parameters
+        ----------
+        X : numpy array of dimensions [2,3,4]
+            The data to be reshaped.
+
+        Returns
+        -------
+        X : numpy ndarray
+            The same array
+        """
+        return X
+
+    def _reshape(self, X):
+        """Recontruct epochs to get a n_trials * n_chan * n_time
+
+        Parameters
+        ----------
+        X : numpy array of dimensions [2,3,4]
```

Comment:
but yes, the output should be 3 dimensional so as to pipeline multiple epochs transformers without having the necessity to initialize them. The vectorizing is then only applied when necessary, i.e. when the following estimator expect a 2D X.

e.g.
```
make_pipeline(
     SpatialFilter(PCA(), n_chan=64),
     SpatialFilter(ICA()), # it can handle 3D X and we don't have to track the number of component specified in the previous step (PCA followed by ICA would be silly but it's for the sake of the example)
     Vectorizer(),  # Takes n_dim X to 2 dimensional X
     LogistitcRegression(),  # expect 2D X
)